pkg/basicstation/cups/server.go

```diff
@@ -0,0 +1,283 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cups
+
+import (
+	"context"
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"net"
+	"net/url"
+	"strings"
+
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/pkg/component"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/rpcmetadata"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+	"go.thethings.network/lorawan-stack/pkg/web"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// ServerConfig is the configuration of the CUPS server.
+type ServerConfig struct {
+	ExplicitEnable  bool `name:"require-explicit-enable" description:"Require gateways to explicitly enable CUPS"`
+	RegisterUnknown struct {
+		Type   string `name:"account-type" description:"Type of account to register unknown gateways to (user|organization)"`
+		ID     string `name:"id" description:"ID of the account to register unknown gateways to"`
+		APIKey string `name:"api-key" description:"API Key to use for unknown gateway registration"`
+	} `name:"owner-for-unknown"`
+	AllowCUPSURIUpdate bool `name:"allow-cups-uri-update" description:"Allow CUPS URI updates"`
```

Comment:
Without this a handover wouldn't be possible right? So, why is it optional?

Reply:
Correct. Only the "boot" CUPS updates the CUPS URI. Most of our deployments will not be a "boot" CUPS, so we don't allow CUPS URI updates by default.

Reply:
I'm not so sure of that. So for a CUPS handover between ttn and `x` network server, we need to rely on the boot cups (Ex: Semtech). 
This shouldn't be the case for TTI Gateways.

Reply:
Nevertheless, most CUPS servers won't do handover

Reply:
The factory default CUPS would be TTI operated for TTIG and TTOG. We can support any LNS. We will probably not allow CUPS-level handover for those gateways, hence this option.